### PR TITLE
Start building Native AOT CoreLib

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
This brings sources over from the CoreRT repo.

One change I'm doing to the files that are brought over is that I'm unifying the interop with what's in Common. Bye bye `api-ms-win-`.

I tested this by dropping the produced CoreLib into CoreRT binary tree on Windows and running the CoreRT simple tests.

TODO:

- [x] Fix Unix build breaks. I assume there are some
- [x] Find out how to properly disable analyzers. I don't care about whitespace at the end of the line in files nobody touched for 5 years. `<EnableAnalyzers>false</EnableAnalyzers>` doesn't work.
- [x] Factor out DefineConstants to directory.build.props at the root of nativeaot. I don't think we want to follow what CoreCLR's CoreLib is doing.
- [ ] Untangle `GCMemoryInfo` where `Common` has definition for datastructures that share CoreCLR's VMs native layout. Mono avoided the problem by PNSEing the API.
- [ ] Shared portion of Enum has code that we likely don't want use.
- [x] EventProvider deleted ES_BUILD_PN that we need to restore.

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
Turns out we don't need ES_BUILD_PN back to build EventSource. The macro still protected from inclusion of some questionable code in EventSource, such as some odd piece of code calling GetMethodBody and interrogating bytes in it to look for patterns. We can fix those as we get bug reports.

File: src/coreclr/src/nativeaot/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+  <!-- Note that various places in SPCL depend on this resource name i.e. TplEventSource -->
+  <ItemGroup Label="Embedded Resources">
+    <EmbeddedResource Include="src\Resources\Strings.resx">
+      <GenerateSource>true</GenerateSource>
+      <ClassName>System.SR</ClassName>
+    </EmbeddedResource>
+  </ItemGroup>
+
+  <PropertyGroup>
+    <DefineConstants>SYSTEM_PRIVATE_CORELIB;ES_BUILD_PN;ES_SESSION_INFO;$(DefineConstants)</DefineConstants>
```
Comment (jkotas) [inline src/coreclr/src/nativeaot/System.Private.CoreLib/System.Private.CoreLib.csproj:12]:
```suggestion
    <DefineConstants>SYSTEM_PRIVATE_CORELIB;$(DefineConstants)</DefineConstants>
```

File: src/coreclr/src/nativeaot/Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    <!-- This prevents the default MsBuild targets from referencing System.Core.dll -->
+    <AddAdditionalExplicitAssemblyReferences>false</AddAdditionalExplicitAssemblyReferences>
+    <!-- These prevent the default MsBuild targets from referencing System.dll and mscorlib.dll -->
+    <ExcludeMscorlibFacade>true</ExcludeMscorlibFacade>
+    <NoStdLib>true</NoStdLib>
+    <NoCompilerStandardLib>true</NoCompilerStandardLib>
+    <SubsystemVersion>6.00</SubsystemVersion>
+    <RuntimeMetadataVersion>v4.0.30319</RuntimeMetadataVersion>
+    <Utf8Output>true</Utf8Output>
+    <HighEntropyVA>true</HighEntropyVA>
```
Comment (jkotas) [inline src/coreclr/src/nativeaot/Directory.Build.props:31]: I suspect  many of these properties are not needed anymore. We can clean it up later.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@
 using System.Globalization;
 using System.Numerics;
 using System.Runtime.InteropServices;
-#if (CORECLR || MONO) && TARGET_WINDOWS
+#if (CORECLR || MONO || CORERT) && TARGET_WINDOWS
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs:15]:
```suggestion
#if !ES_BUILD_STANDALONE && TARGET_WINDOWS
```

File: src/coreclr/src/nativeaot/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 1 above)
Comment (jkotas) [reply src/coreclr/src/nativeaot/System.Private.CoreLib/System.Private.CoreLib.csproj:12]: `ES_SESSION_INFO` can be deleted from the sources too. The ifdef can be just `#if !ES_BUILD_STANDALONE && TARGET_WINDOWS`

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs (hunk 3 above)
Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs:15]: How do we want to approach these cleanups? Do we want to accumulate them in the branch, or do the cleanup in the runtime repo?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs:15]: I would accumulate one liners like this one in the branch and flush them out once in a while.

**Comment by jkotas** [comment]
I think this is ready to merge - we can open issues on the remaining checkboxes.

**Comment by MichalStrehovsky** [comment]
> I think this is ready to merge - we can open issues on the remaining checkboxes.

yup, I'll file the issue & merge

**Pull request merged**